8269429: Linux: Only the last APPLICATION_MODAL window behaves correctly

tsayao opened the pull request:
The PR approach is to set  `gtk_window_set_keep_above` to true on APPLICATION_MODAL windows, so they will not stay behind non APPLICATION_MODAL windows.

This is passed on WindowStage.java:198 as a mask.

The weird thing is that `_enterModal()` is never called. This seems the right function to be called for `APPLICATION_MODAL`, as `_enterModalWithWindow` fits for `WINDOW_MODAL`.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8269429](https://bugs.openjdk.java.net/browse/JDK-8269429): Linux: Only the last APPLICATION_MODAL window behaves correctly


### Reviewers
 * [Pankaj Bansal](https://openjdk.java.net/census#pbansal) (@pankaj-bansal - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/551/head:pull/551` \
`$ git checkout pull/551`

Update a local copy of the PR: \
`$ git checkout pull/551` \
`$ git pull https://git.openjdk.java.net/jfx pull/551/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 551`

View PR using the GUI difftool: \
`$ git pr show -t 551`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/551.diff">https://git.openjdk.java.net/jfx/pull/551.diff</a>

</details>

tsayao commented:
It's odd that glass `Window.java` defines `enterModal` but it seems to be never called. Modality is handled by calling `setEnabled(false)` on other windows.

I would refactor that some day :)

tsayao commented:
It's working as expected:

If there's two APPLICATION_MODAL (*1) and (*2) and a WINDOW_MODAL (*3) as a child of the primary stage the order of the windows on the screen will be (*2) (*1) (*3).

Will run the tests and post here.

kevinrushforth commented:
/reviewers 2

kevinrushforth reviewed (commented):
I tried the test program attached to the JBS bug, and I'm not seeing the problem. Are there any special steps I need to take to see the bug? I also left a question about `WINDOW_MODAL` vs `APPLICATION_MODAL` windows.

Regarding the `{enter/exit/is}Modal` methods in `Window.java`, it looks like they stubs that are never called. Modality is handled by the toolkit using (as you found out) only the `setEnabled` method.

Finally, you have a typo in the title of both this PR and the JBS bug: `corretly` --> `correctly`

tsayao commented on modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp:
```diff
@@ -735,6 +736,9 @@ WindowContextTop::WindowContextTop(jobject _jwindow, WindowContext* _owner, long
         }
+    } else if (modal) {
+        //this is used for APPLICATION_MODAL
+        gtk_window_set_keep_above(GTK_WINDOW(gtk_widget), TRUE);
```
Yes. I have refactored it.

tsayao commented:
> I tried the test program attached to the JBS bug, and I'm not seeing the problem. Are there any special steps I need to take to see the bug? I also left a question about `WINDOW_MODAL` vs `APPLICATION_MODAL` windows.
> 
> Regarding the `{enter/exit/is}Modal` methods in `Window.java`, it looks like they stubs that are never called. Modality is handled by the toolkit using (as you found out) only the `setEnabled` method.
> 
> Finally, you have a typo in the title of both this PR and the JBS bug: `corretly` --> `correctly`

- Added steps to reproduce on JBS. Basically the window manager will let you click the window in the back (primary stage in the example) and the APPLICATION_MODAL will go behind it, still blocking events;
- Fixed the typo.

tsayao commented:
![image](https://user-images.githubusercontent.com/30704286/123880825-d1129d80-d919-11eb-8e49-07f18da72dd1.png)

tsayao commented:
The first approach was to make the window not clickable or not "activable" when `setEnabled(false)` but I did not find a solution for that on Gdk or Gtk. Gtk has `gtk_widget_set_sensitive` but that did not work (probably only if using gtk signals). I did make other attempts with no success.

pankaj-bansal commented:
I tried this on Ubuntu 20.04 and I see that the patch does solve the issue by bringing the Alert window above the primary window

kevinrushforth commented on modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp:
```diff
@@ -1425,9 +1430,10 @@ void WindowContextTop::restack(bool restack) {
         if (parent) {
             gtk_window_set_transient_for(GTK_WINDOW(gtk_widget), parent->get_gtk_window());
+        } else {
+            gtk_window_set_keep_above(GTK_WINDOW(gtk_widget), modal ? TRUE : FALSE);
```
The test of `modal` is unnecessary here, since this is all within an `if (modal)` block.

tsayao replied:
Done.

kevinrushforth reviewed (commented):
This looks fine to me. I'll test it on a couple different systems. I left one comment inline.

kevinrushforth commented:
I verified that this fixes the problem, but it also changes the behavior of `APPLICATION_MODAL` dialogs on Linux to be "Always on Top" -- not just on top of other application windows (which is fine), but on top of all other windows an application might create, which is not desirable as a default behavior. This also makes it behave differently than Mac or Windows.

I think we should look for a solution that doesn't rely on forcing a dialog to be on top of other (no app) windows.

tsayao commented:
Will try to find a solution to disable events on disabled windows `setEnable(false)`

tsayao commented:
This is also fixed by #581.
